docs/ibc/custom.md

```diff
@@ -142,6 +142,16 @@ OnChanCloseConfirm(
 }
 ```
 
+#### Channel Handshake Version Negotiation
+
+Application modules are exepcted to verify versioning used during the channel handshake procedure.
+
+* `ChanOpenInit` callback should verify that the `MsgChanOpenInit.Version` is valid
+* `ChanOpenTry` callback should verify that the `MsgChanOpenTry.Version` is valid and that `MsgChanOpenTry.CounterpartyVersion` is valid.
+* `ChanOpenAck` callback should verify that the `MsgChanOpenAck.CounterpartyVersion` is valid and supported.
+
+Versions must be strings but can implement any versioning structure.
```

Comment:
My understanding was that the callbacks were used to determine if the module supported the version at each step in the negotiation, but the negotiation itself was standardized and handled by IBC just as with connections (just using callbacks to find which versions are supported/valid).

Seems useful to coalesce around a standard for how IBC apps specify version strings and how they are negotiated.

cc: @cwgoes

Reply:
This would require changing the channel handshake API, which is why I didn't want to attempt this approach. ChanOpenInit and ChanOpenTry accept a single string not a set of strings as the args so no selection process could happen outside of picking a subset of features. 

In all likelyhood, the user creating the channels (relayer) will be calling all 4 handshake procedure steps and there won't be any "negotiation" occurring but rather a single selection which occurs before OpenInit is even called. I think just verification and the possibility for version negotiation on the application side is sufficient for 1.0. It is not entirely clear how often or how complex applications plan to use versioning (if at all).

Reply:
In short, I don't think standardizing application version negotiation is worth the overhead/complexity in combination with the added restrictions on how applications of how to represent versions. Especially since I believe the connection version negotiation could change with feedback from 1.0 usage

Reply:
The question here is as to whether the version is standardised at the protocol layer as a string, or as structured data, right? At the connection layer, versions are also just "string" in the protocol spec, but the SDK has a particular standard. It may be worth standardising this at the specification layer; if the SDK uses a particular format people will likely adopt it anyways, originally I was hesitant to do that but I'm starting to change my mind, I'm not sure what the "string" abstraction really buys us.

Reply:
I think there is two questions here:
* versions as a string or structured data
* version selection occurring in the channel handshake

> As mentioned above, I think the selection process should be done by the channel handler, just using the module-specific callbacks to determine whether to accept/reject a version string at each step in the negotiation.

I was primarily responding to this line on the comment down below (I responded in the incorrect location). 

I agree that creating a structured data is probably ideal at the connection layer and maybe even at the channel layer. But I disagree that that selection at the channel layer is particularly beneficial at this current time.

Reply:
If we add a structure version to the specification, the channel handshake callbacks should use the structured version directly, definitely. (does that answer your question?)